### PR TITLE
Breaking: Add tests for Client

### PR DESCRIPTION
**Pull request by vogdb**
Acceptance tests on Travis?

### CODE REVIEW

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by vogdb** [comment]
@tomaszdurka please review

File: client/index.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  Client.prototype.open = function() {
+    clearTimeout(this._recopenTimeout);
+    this._recopenTimeout = null;
+
+    this._sockJS = new SockJS(this._url);
+    this._sockJS.onopen = this._onopen.bind(this);
+    this._sockJS.onclose = this._recopen.bind(this);
+  };
+
+  Client.prototype._recopen = function() {
```
Comment (tomaszdurka) [inline client/index.js]: Is this meant to be "**rec**open"?

**Comment by tomaszdurka** [comment]
This is big refactoring for the socket-redis project. I have some other priorities now so didn't look into this. Maybe we can have a pairing-call regarding this once I'm bit more free. Thanks!

**Comment by vogdb** [comment]
ok, just poke me whenever you are free.

**Comment by vogdb** [comment]
@tomaszdurka please review

File: client/index.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    this.onclose.call(this);
+  };
+
+  /**
+   * @param {String} channel
+   * @param {Number} [start]
+   * @param {Object} [data]
+   * @param {Function} onmessage fn(data)
+   */
+  Client.prototype.subscribe = function(channel, start, data, onmessage) {
```
Comment (tomaszdurka) [inline client/index.js]: Bit strange signature (optional params before onmessage), that's probably due callback multiline constructor.
What about making `onmessage` optional too? I guess you might want only to subscribe and do nothing on message?

**Comment by tomaszdurka** [comment]
This looks good to me (short notes), it would be great to test this out with working product locally before merging. Did you have time to confront it with SK?

File: client/index.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+   * @param {String} channel
+   */
+  Client.prototype.unsubscribe = function(channel) {
+    if (this._subscribes[channel]) {
+      delete this._subscribes[channel];
+    }
+    if (this._sockJS.readyState === SockJS.OPEN) {
+      this._sockJS.send(JSON.stringify({event: 'unsubscribe', data: {channel: channel}}));
```
Comment (tomaszdurka) [inline client/index.js]: Should we queue event for sockjs?

File: client/index.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+  };
+
+  /**
+   * @param {Object} data
+   */
+  Client.prototype.send = function(data) {
+    this._sockJS.send(JSON.stringify({event: 'message', data: {data: data}}));
```
Comment (tomaszdurka) [inline client/index.js]: Should we queue sends for sockjs until it's open or it does it on its own?

File: client/index.js (hunk 2 above)
Comment (vogdb) [inline client/index.js]: Well, I preserved the signature for backward compatibility. `onmessage` can be optional but it does not make any sense. It was optional before but due to mistake because when omitted it produced an exception.
I would propose to change the signature then and keep it as required. Your thoughts?

File: client/index.js (hunk 4 above)
Comment (vogdb) [inline client/index.js]: we may but it should be done in different PR cause this feature hasn't been implemented before.

**Comment by vogdb** [comment]
@tomaszdurka please look at my answers to your notes

File: client/index.js (hunk 2 above)
Comment (tomaszdurka) [inline client/index.js]: We can do either for me. I think subscribe-only behaviour could be useful for some. More like presence channel. Your call.

File: client/index.js (hunk 4 above)
Comment (tomaszdurka) [inline client/index.js]: Agree. Open if you think it's necessary.

**Comment by vogdb** [comment]
> Did you have time to confront it with SK?

The changes are breaking, so it won't work with the current SK.

**Comment by vogdb** [comment]
I want to propose the next. We merge this. I rework server side part entirely. We release those changes as `3.0.0`. @tomaszdurka ?

**Comment by vogdb** [comment]
or to make things more smooth and calm. We merge this PR to a new branch on this repo like `new-socket-redis`. I rework server and merge it to `new-socket-redis` too. After that we do a PR from `new-socket-redis` to `master`, test everything carefully and merge.

**Comment by tomaszdurka** [comment]
I reviewed new code, but not so much in comparison with old. I understand this communicates the same way with server.

**Comment by vogdb** [comment]
@tomaszdurka what about my questions?
> I want to propose the next. We merge this. I rework server side part entirely. We release those changes as `3.0.0`. @tomaszdurka ?

> or to make things more smooth and calm. We merge this PR to a new branch on this repo like `new-socket-redis`. I rework server and merge it to `new-socket-redis` too. After that we do a PR from `new-socket-redis` to `master`, test everything carefully and merge.

What do you choose?

**Comment by tomaszdurka** [comment]
I counter-*asked* :) Can't we merge client changes without server changes?

**Comment by tomaszdurka** [comment]
I understand we can merge it, so my question before was could we test new client (with still old server) on sk product.

**Comment by vogdb** [comment]
> my question before was could we test new client (with still old server) on sk product

I tested. It works but requires some changes. You was mentioned in this issue.

> Can't we merge client changes without server changes?

We can merge. What version should we pick then? Client breaks compatibility. Should we release with new client only?

**Comment by tomaszdurka** [comment]
Ok it all makes sense, was confused by:

>  > Did you have time to confront it with SK?

> The changes are breaking, so it won't work with the current SK.

I understood that you could not migrate the new client because it's breaking on sk with... yeah what? :o
Ok, so new client has been tested on SK, works fine, let's merge this. We don't need to release it yet indeed. Just wanted to make sure current master is fine.

**Comment by vogdb** [comment]
> I understood that you could not migrate the new client because it's breaking on sk with... yeah what?

It does not work with the current SK. SK needs to be adjusted. The required changes are mentioned in the SK issue that is responsible for those changes.

**Comment by tomaszdurka** [comment]
This looks good to me.

**Comment by vogdb** [comment]
@tomaszdurka unfortunately I had to those changes. Please review them.

File: client/index.js [hunk 5]
```diff
@@ -25,6 +25,7 @@ Client.prototype.open = function() {
   clearTimeout(this._reopenTimeout);
   this._reopenTimeout = null;
 
+  SockJS = require('sockjs-client');
```
Comment (tomaszdurka) [inline client/index.js]: Hmm, otherwise you cannot build project properly or what's the story?

**Comment by tomaszdurka** [comment]
Well changes are ok.

File: client/index.js (hunk 5 above)
Comment (vogdb) [inline client/index.js]: The problem comes from browserify. It requires to have `window.SockJS` before this script which can not be done for our usecase. In our setup `window.SockJS` is defined later than `window.SocketRedis`. So when we execute `socket-redis.js` `window.SockJS` is not yet defined.

Comment (tomaszdurka) [inline client/index.js]: ok then

**Comment by vogdb** [comment]
Closing in favor of #84
